src/storage/src/source/kafka.rs

```diff
@@ -104,382 +98,423 @@ pub struct KafkaOffsetCommiter {
+                            if let Some((msg, time, diff)) = reader.handle_message(message, ts) {
+                                let pid = time.partition().unwrap();
+                                let part_cap = &reader.partition_capabilities[pid];
+                                data_output.give(part_cap, (Ok(msg), time, diff)).await;
+                            }
+                        }
+                    }
+                }
 
-            match message {
-                Ok(Some((message, ts))) => {
-                    next_message = self.handle_message(Ok(message), ts);
+                reader.update_stats();
+
+                // Take the consumers temporarily to get around borrow checker errors
+                let mut consumers = std::mem::take(&mut reader.partition_consumers);
+                for consumer in consumers.iter_mut() {
+                    while let Some(message) = consumer.get_next_message().transpose() {
+                        match message.map(|(msg, ts)| reader.handle_message(msg, ts)) {
+                            Ok(Some((msg, time, diff))) => {
+                                let pid = time.partition().unwrap();
+                                let part_cap = &reader.partition_capabilities[pid];
+                                data_output.give(part_cap, (Ok(msg), time, diff)).await;
+                            }
+                            Ok(None) => continue,
+                            Err(err) => {
+                                let pid = consumer.pid();
+                                let last_offset = reader
+                                    .last_offsets
+                                    .get(&pid)
+                                    .expect("partition known to be installed");
+
+                                let status = HealthStatus::StalledWithError {
+                                    error: format!(
+                                        "error consuming from source: {} topic: {topic}: partition:\
+                                        {pid} last processed offset: {last_offset} : {err}",
+                                        config.name
+                                    ),
+                                    hint: None,
+                                };
+                                health_output.give(&health_cap, status.into()).await;
+                            }
+                        }
+                    }
                 }
-                Err(error) => {
-                    next_message = NextMessage::Ready(SourceMessageType::status(
-                        HealthStatus::StalledWithError { error, hint: None },
-                    ))
+                // We can now put them back
+                assert!(reader.partition_consumers.is_empty());
+                reader.partition_consumers = consumers;
+
+                for (pid, last_offset) in reader.last_offsets.iter() {
```

Comment:
is `last_offsets` updated when we `poll`?

Reply:
It's updated in `get_next_message()`